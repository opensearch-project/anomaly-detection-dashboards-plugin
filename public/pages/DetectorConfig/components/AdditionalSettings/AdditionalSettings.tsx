/*
 * SPDX-License-Identifier: Apache-2.0
 *
 * The OpenSearch Contributors require contributions made to
 * this file be licensed under the Apache-2.0 license or a
 * compatible open source license.
 *
 * Modifications Copyright OpenSearch Contributors. See
 * GitHub history for details.
 */

import React, { useState } from 'react';
import { get, isEmpty } from 'lodash';
import {
  EuiBasicTable,
  EuiButtonEmpty,
  EuiOverlayMask
} from '@elastic/eui';
import ContentPanel from '../../../../components/ContentPanel/ContentPanel';
import { convertToCategoryFieldString } from '../../../utils/anomalyResultUtils';
import { SuppressionRulesModal } from '../../../ReviewAndCreate/components/AdditionalSettings/SuppressionRulesModal';

interface AdditionalSettingsProps {
  shingleSize: number;
  categoryField: string[];
  imputationMethod: string;
  customValues: string[];
  suppressionRules: string[];
}

export function AdditionalSettings(props: AdditionalSettingsProps) {
  const renderCustomValues = (customValues: string[]) => (
    <div>
      {customValues.length > 0 ? (
        customValues.map((value, index) => <p key={index}>{value}</p>)
      ) : (
        <p>-</p>
      )}
    </div>
  );

  const [isModalVisible, setIsModalVisible] = useState(false);
  const [modalContent, setModalContent] = useState<string[]>([]);

  const closeModal = () => setIsModalVisible(false);

  const showRulesInModal = (rules: string[]) => {
    setModalContent(rules);
    setIsModalVisible(true);
  };

  const renderSuppressionRules = (suppressionRules: string[]) => (
    <div>
      {suppressionRules.length > 0 ? (
        <EuiButtonEmpty size="s" onClick={() => showRulesInModal(suppressionRules)}>
          {suppressionRules.length} rules
        </EuiButtonEmpty>
      ) : (
        <p>-</p>
      )}
    </div>
  );

  const tableItems = [
    {
      categoryField: isEmpty(get(props, 'categoryField', []))
        ? '-'
        : convertToCategoryFieldString(props.categoryField, ', '),
      shingleSize: props.shingleSize,
      imputationMethod: props.imputationMethod,
      customValues: props.customValues,
      suppresionRules: props.suppressionRules,
    },
  ];
  const tableColumns = [
    { name: 'Categorical fields', field: 'categoryField' },
    { name: 'Shingle size', field: 'shingleSize' },
    { name: 'Imputation method', field: 'imputationMethod' },
    { name: 'Custom values',
      field: 'customValues',
      render: (customValues: string[]) => renderCustomValues(customValues), // Use a custom render function
     },
     { name: 'Suppression rules',
      field: 'suppresionRules',
      render: (suppresionRules: string[]) => renderSuppressionRules(suppresionRules), // Use a custom render function
     },
  ];
  return (
    <ContentPanel title="Additional settings" titleSize="s">
      <EuiBasicTable
        className="header-single-value-euiBasicTable"
        items={tableItems}
        columns={tableColumns}
      />
      {isModalVisible && (
        <EuiOverlayMask>
          <SuppressionRulesModal onClose={closeModal} rules={modalContent} />
        </EuiOverlayMask>
      )}
    </ContentPanel>
  );
}
