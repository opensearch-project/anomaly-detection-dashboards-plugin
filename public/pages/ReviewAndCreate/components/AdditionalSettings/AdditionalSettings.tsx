/*
 * SPDX-License-Identifier: Apache-2.0
 *
 * The OpenSearch Contributors require contributions made to
 * this file be licensed under the Apache-2.0 license or a
 * compatible open source license.
 *
 * Modifications Copyright OpenSearch Contributors. See
 * GitHub history for details.
 */

import React, { useState } from 'react';
import { get } from 'lodash';
import {
  EuiBasicTable,
  EuiButtonEmpty,
  EuiOverlayMask
} from '@elastic/eui';
import ContentPanel from '../../../../components/ContentPanel/ContentPanel';
import { SuppressionRulesModal } from './SuppressionRulesModal';

interface AdditionalSettingsProps {
  shingleSize: number;
  categoryField: string[];
  imputationMethod: string;
  customValues: string[];
  suppressionRules: string[];
}

export function AdditionalSettings(props: AdditionalSettingsProps) {
  const renderCustomValues = (customValues: string[]) => (
    <div>
      {customValues.length > 0 ? (
        customValues.map((value, index) => <p key={index}>{value}</p>)
      ) : (
        <p>-</p>
      )}
    </div>
  );

  const [isModalVisible, setIsModalVisible] = useState(false);
  const [modalContent, setModalContent] = useState<string[]>([]);

  const closeModal = () => setIsModalVisible(false);

  const showRulesInModal = (rules: string[]) => {
    setModalContent(rules);
    setIsModalVisible(true);
  };

  const renderSuppressionRules = (suppressionRules: string[]) => (
    <div>
      {suppressionRules.length > 0 ? (
        <EuiButtonEmpty size="s" onClick={() => showRulesInModal(suppressionRules)}>
          {suppressionRules.length} rules
        </EuiButtonEmpty>
      ) : (
        <p>-</p>
      )}
    </div>
  );

  const tableItems = [
    {
      categoryField: get(props, 'categoryField.0', '-'),
      shingleSize: props.shingleSize,
      imputationMethod: props.imputationMethod,
      customValues: props.customValues,
      suppresionRules: props.suppressionRules,
    },
  ];
  const tableColumns = [
    { name: 'Category field', field: 'categoryField' },
    { name: 'Shingle size', field: 'shingleSize' },
    { name: 'Imputation method', field: 'imputationMethod' },
    { name: 'Custom values',
      field: 'customValues',
      render: (customValues: string[]) => renderCustomValues(customValues), // Use a custom render function
    },
    { name: 'Suppression rules',
      field: 'suppresionRules',
      render: (suppressionRules: string[]) => renderSuppressionRules(suppressionRules), // Use a custom render function
    },
  ];
  return (
    <ContentPanel title="Additional settings" titleSize="s">
    <EuiBasicTable
      className="header-single-value-euiBasicTable"
      items={tableItems}
      columns={tableColumns}
    />
    {isModalVisible && (
        <EuiOverlayMask>
          <SuppressionRulesModal onClose={closeModal} rules={modalContent} />
        </EuiOverlayMask>
      )}
    </ContentPanel>
  );
}
