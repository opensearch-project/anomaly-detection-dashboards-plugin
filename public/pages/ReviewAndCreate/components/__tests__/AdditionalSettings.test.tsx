/*
 * SPDX-License-Identifier: Apache-2.0
 *
 * The OpenSearch Contributors require contributions made to
 * this file be licensed under the Apache-2.0 license or a
 * compatible open source license.
 *
 * Modifications Copyright OpenSearch Contributors. See
 * GitHub history for details.
 */

import React from 'react';
import { render, fireEvent, screen, waitFor } from '@testing-library/react';
import { AdditionalSettings } from '../AdditionalSettings/AdditionalSettings';

import { Formik } from 'formik';

describe('<AdditionalSettings /> spec', () => {
  test('renders the component with high cardinality disabled', () => {
    const { container, getByText, getAllByText, queryByRole } = render(
      <Formik initialValues={{ detectorName: '' }} onSubmit={jest.fn()}>
        {() => (
          <div>
            <AdditionalSettings
              categoryField={[]}
              shingleSize={8}
              imputationMethod="Ignore"
              customValues={[]}
              suppressionRules={[]}
            />
          </div>
        )}
      </Formik>
    );
    expect(container.firstChild).toMatchSnapshot();
    getAllByText('Category field');
    getAllByText('Shingle size');
    getByText('8');
    getByText('Ignore');

    // Assert that multiple elements with the text '-' are present
    const dashElements = getAllByText('-');
    expect(dashElements.length).toBeGreaterThan(1); // Checks that more than one '-' is found

    // Check that the 'Suppression rules' title is present
    // Assert that multiple elements with the text '-' are present
    const ruleElements = getAllByText('Suppression rules');
    expect(ruleElements.length).toBeGreaterThan(1); // one is table cell title, another is the button

    // Use queryByRole to check that the button link is not present
    const button = screen.queryByRole('button', { name: '0 rules' });
    expect(button).toBeNull();
  });
  test('renders the component with high cardinality enabled', async () => {
    const { container, getByText, getAllByText, getByRole, queryByRole } =
      render(
        <Formik initialValues={{ detectorName: '' }} onSubmit={jest.fn()}>
          {() => (
            <div>
              <AdditionalSettings
                categoryField={['test_field']}
                shingleSize={8}
                imputationMethod="Custom"
                customValues={['denyMax:5', 'denySum:10']}
                suppressionRules={[
                  "Ignore anomalies for feature 'CPU Usage' with no more than 5 above expected value.",
                  "Ignore anomalies for feature 'Memory Usage' with no more than 10% below expected value.",
                ]}
              />
            </div>
          )}
        </Formik>
      );
    expect(container.firstChild).toMatchSnapshot();
    getAllByText('Category field');
    getAllByText('Shingle size');
    getByText('test_field');
    getByText('8');
    getByText('Custom');
    // Check for the custom values
    getByText('denyMax:5');
    getByText('denySum:10');

    // Check for the suppression rules button link
    const button = getByRole('button', { name: '2 rules' });
    expect(button).toBeInTheDocument();

    // Click the button to open the modal
    fireEvent.click(button);

    // Wait for the modal to appear and check for its content
    await waitFor(() => {
      expect(screen.getByRole('dialog')).toBeInTheDocument(); // Ensure modal is opened
    });

    getByText('Suppression Rules'); // Modal header
    getByText(
      "Ignore anomalies for feature 'CPU Usage' with no more than 5 above expected value."
    );
    getByText(
      "Ignore anomalies for feature 'Memory Usage' with no more than 10% below expected value."
    );

    // Close the modal by clicking the close button (X)
    // Close the modal by clicking the close button (X)
    const closeButton = getByRole('button', {
      name: 'Closes this modal window',
    });
    fireEvent.click(closeButton);

    // Ensure the modal is closed
    await waitFor(() => {
      expect(queryByRole('dialog')).toBeNull();
    });
  });
});
