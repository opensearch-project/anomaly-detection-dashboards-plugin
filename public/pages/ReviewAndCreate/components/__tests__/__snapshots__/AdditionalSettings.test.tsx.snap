// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<AdditionalSettings /> spec renders the component with high cardinality disabled 1`] = `
<div>
  <div
    class="euiPanel euiPanel--paddingMedium euiPanel--borderRadiusMedium euiPanel--plain euiPanel--hasShadow"
    style="padding: 20px;"
  >
    <div
      class="euiFlexGroup euiFlexGroup--gutterLarge euiFlexGroup--alignItemsCenter euiFlexGroup--justifyContentSpaceBetween euiFlexGroup--directionRow euiFlexGroup--responsive"
      style="padding: 0px;"
    >
      <div
        class="euiFlexItem"
      >
        <h3
          class="euiTitle euiTitle--small"
          data-test-subj="contentPanelTitle"
        >
          Additional settings
        </h3>
        <div
          class="euiFlexGroup euiFlexGroup--gutterLarge euiFlexGroup--directionRow euiFlexGroup--responsive"
        >
          <div
            class="euiFlexItem content-panel-subTitle"
            style="line-height: normal; max-width: 75%;"
          />
        </div>
      </div>
      <div
        class="euiFlexItem euiFlexItem--flexGrowZero"
      >
        <div
          class="euiFlexGroup euiFlexGroup--gutterMedium euiFlexGroup--alignItemsCenter euiFlexGroup--justifyContentSpaceBetween euiFlexGroup--directionRow euiFlexGroup--responsive"
        >
          <div
            class="euiFlexItem"
          />
        </div>
      </div>
    </div>
    <div>
      <hr
        class="euiHorizontalRule euiHorizontalRule--full euiHorizontalRule--marginSmall"
      />
      <div
        style="padding: 10px 0px;"
      >
        <div
          class="euiBasicTable header-single-value-euiBasicTable"
        >
          <div>
            <div
              class="euiTableHeaderMobile"
            >
              <div
                class="euiFlexGroup euiFlexGroup--gutterLarge euiFlexGroup--alignItemsBaseline euiFlexGroup--justifyContentSpaceBetween euiFlexGroup--directionRow"
              >
                <div
                  class="euiFlexItem euiFlexItem--flexGrowZero"
                />
                <div
                  class="euiFlexItem euiFlexItem--flexGrowZero"
                />
              </div>
            </div>
            <table
              class="euiTable euiTable--responsive"
              id="random_html_id"
              tabindex="-1"
            >
              <caption
                class="euiScreenReaderOnly euiTableCaption"
              />
              <thead>
                <tr>
                  <th
                    class="euiTableHeaderCell"
                    data-test-subj="tableHeaderCell_categoryField_0"
                    role="columnheader"
                    scope="col"
                  >
                    <span
                      class="euiTableCellContent"
                    >
                      <span
                        class="euiTableCellContent__text"
                        title="Category field"
                      >
                        Category field
                      </span>
                    </span>
                  </th>
                  <th
                    class="euiTableHeaderCell"
                    data-test-subj="tableHeaderCell_shingleSize_1"
                    role="columnheader"
                    scope="col"
                  >
                    <span
                      class="euiTableCellContent"
                    >
                      <span
                        class="euiTableCellContent__text"
                        title="Shingle size"
                      >
                        Shingle size
                      </span>
                    </span>
                  </th>
                  <th
                    class="euiTableHeaderCell"
                    data-test-subj="tableHeaderCell_imputationMethod_2"
                    role="columnheader"
                    scope="col"
                  >
                    <span
                      class="euiTableCellContent"
                    >
                      <span
                        class="euiTableCellContent__text"
                        title="Imputation method"
                      >
                        Imputation method
                      </span>
                    </span>
                  </th>
                  <th
                    class="euiTableHeaderCell"
                    data-test-subj="tableHeaderCell_customValues_3"
                    role="columnheader"
                    scope="col"
                  >
                    <span
                      class="euiTableCellContent"
                    >
                      <span
                        class="euiTableCellContent__text"
                        title="Custom values"
                      >
                        Custom values
                      </span>
                    </span>
                  </th>
                  <th
                    class="euiTableHeaderCell"
                    data-test-subj="tableHeaderCell_suppresionRules_4"
                    role="columnheader"
                    scope="col"
                  >
                    <span
                      class="euiTableCellContent"
                    >
                      <span
                        class="euiTableCellContent__text"
                        title="Suppression rules"
                      >
                        Suppression rules
                      </span>
                    </span>
                  </th>
                </tr>
              </thead>
              <tbody>
                <tr
                  class="euiTableRow"
                >
                  <td
                    class="euiTableRowCell"
                  >
                    <div
                      class="euiTableRowCell__mobileHeader euiTableRowCell--hideForDesktop"
                    >
                      Category field
                    </div>
                    <div
                      class="euiTableCellContent"
                    >
                      <span
                        class="euiTableCellContent__text"
                      >
                        -
                      </span>
                    </div>
                  </td>
                  <td
                    class="euiTableRowCell"
                  >
                    <div
                      class="euiTableRowCell__mobileHeader euiTableRowCell--hideForDesktop"
                    >
                      Shingle size
                    </div>
                    <div
                      class="euiTableCellContent"
                    >
                      <span
                        class="euiTableCellContent__text"
                      >
                        8
                      </span>
                    </div>
                  </td>
                  <td
                    class="euiTableRowCell"
                  >
                    <div
                      class="euiTableRowCell__mobileHeader euiTableRowCell--hideForDesktop"
                    >
                      Imputation method
                    </div>
                    <div
                      class="euiTableCellContent"
                    >
                      <span
                        class="euiTableCellContent__text"
                      >
                        Ignore
                      </span>
                    </div>
                  </td>
                  <td
                    class="euiTableRowCell"
                  >
                    <div
                      class="euiTableRowCell__mobileHeader euiTableRowCell--hideForDesktop"
                    >
                      Custom values
                    </div>
                    <div
                      class="euiTableCellContent euiTableCellContent--overflowingContent"
                    >
                      <div
                        class=""
                      >
                        <p>
                          -
                        </p>
                      </div>
                    </div>
                  </td>
                  <td
                    class="euiTableRowCell"
                  >
                    <div
                      class="euiTableRowCell__mobileHeader euiTableRowCell--hideForDesktop"
                    >
                      Suppression rules
                    </div>
                    <div
                      class="euiTableCellContent euiTableCellContent--overflowingContent"
                    >
                      <div
                        class=""
                      >
                        <p>
                          -
                        </p>
                      </div>
                    </div>
                  </td>
                </tr>
              </tbody>
            </table>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`<AdditionalSettings /> spec renders the component with high cardinality enabled 1`] = `
<div>
  <div
    class="euiPanel euiPanel--paddingMedium euiPanel--borderRadiusMedium euiPanel--plain euiPanel--hasShadow"
    style="padding: 20px;"
  >
    <div
      class="euiFlexGroup euiFlexGroup--gutterLarge euiFlexGroup--alignItemsCenter euiFlexGroup--justifyContentSpaceBetween euiFlexGroup--directionRow euiFlexGroup--responsive"
      style="padding: 0px;"
    >
      <div
        class="euiFlexItem"
      >
        <h3
          class="euiTitle euiTitle--small"
          data-test-subj="contentPanelTitle"
        >
          Additional settings
        </h3>
        <div
          class="euiFlexGroup euiFlexGroup--gutterLarge euiFlexGroup--directionRow euiFlexGroup--responsive"
        >
          <div
            class="euiFlexItem content-panel-subTitle"
            style="line-height: normal; max-width: 75%;"
          />
        </div>
      </div>
      <div
        class="euiFlexItem euiFlexItem--flexGrowZero"
      >
        <div
          class="euiFlexGroup euiFlexGroup--gutterMedium euiFlexGroup--alignItemsCenter euiFlexGroup--justifyContentSpaceBetween euiFlexGroup--directionRow euiFlexGroup--responsive"
        >
          <div
            class="euiFlexItem"
          />
        </div>
      </div>
    </div>
    <div>
      <hr
        class="euiHorizontalRule euiHorizontalRule--full euiHorizontalRule--marginSmall"
      />
      <div
        style="padding: 10px 0px;"
      >
        <div
          class="euiBasicTable header-single-value-euiBasicTable"
        >
          <div>
            <div
              class="euiTableHeaderMobile"
            >
              <div
                class="euiFlexGroup euiFlexGroup--gutterLarge euiFlexGroup--alignItemsBaseline euiFlexGroup--justifyContentSpaceBetween euiFlexGroup--directionRow"
              >
                <div
                  class="euiFlexItem euiFlexItem--flexGrowZero"
                />
                <div
                  class="euiFlexItem euiFlexItem--flexGrowZero"
                />
              </div>
            </div>
            <table
              class="euiTable euiTable--responsive"
              id="random_html_id"
              tabindex="-1"
            >
              <caption
                class="euiScreenReaderOnly euiTableCaption"
              />
              <thead>
                <tr>
                  <th
                    class="euiTableHeaderCell"
                    data-test-subj="tableHeaderCell_categoryField_0"
                    role="columnheader"
                    scope="col"
                  >
                    <span
                      class="euiTableCellContent"
                    >
                      <span
                        class="euiTableCellContent__text"
                        title="Category field"
                      >
                        Category field
                      </span>
                    </span>
                  </th>
                  <th
                    class="euiTableHeaderCell"
                    data-test-subj="tableHeaderCell_shingleSize_1"
                    role="columnheader"
                    scope="col"
                  >
                    <span
                      class="euiTableCellContent"
                    >
                      <span
                        class="euiTableCellContent__text"
                        title="Shingle size"
                      >
                        Shingle size
                      </span>
                    </span>
                  </th>
                  <th
                    class="euiTableHeaderCell"
                    data-test-subj="tableHeaderCell_imputationMethod_2"
                    role="columnheader"
                    scope="col"
                  >
                    <span
                      class="euiTableCellContent"
                    >
                      <span
                        class="euiTableCellContent__text"
                        title="Imputation method"
                      >
                        Imputation method
                      </span>
                    </span>
                  </th>
                  <th
                    class="euiTableHeaderCell"
                    data-test-subj="tableHeaderCell_customValues_3"
                    role="columnheader"
                    scope="col"
                  >
                    <span
                      class="euiTableCellContent"
                    >
                      <span
                        class="euiTableCellContent__text"
                        title="Custom values"
                      >
                        Custom values
                      </span>
                    </span>
                  </th>
                  <th
                    class="euiTableHeaderCell"
                    data-test-subj="tableHeaderCell_suppresionRules_4"
                    role="columnheader"
                    scope="col"
                  >
                    <span
                      class="euiTableCellContent"
                    >
                      <span
                        class="euiTableCellContent__text"
                        title="Suppression rules"
                      >
                        Suppression rules
                      </span>
                    </span>
                  </th>
                </tr>
              </thead>
              <tbody>
                <tr
                  class="euiTableRow"
                >
                  <td
                    class="euiTableRowCell"
                  >
                    <div
                      class="euiTableRowCell__mobileHeader euiTableRowCell--hideForDesktop"
                    >
                      Category field
                    </div>
                    <div
                      class="euiTableCellContent"
                    >
                      <span
                        class="euiTableCellContent__text"
                      >
                        test_field
                      </span>
                    </div>
                  </td>
                  <td
                    class="euiTableRowCell"
                  >
                    <div
                      class="euiTableRowCell__mobileHeader euiTableRowCell--hideForDesktop"
                    >
                      Shingle size
                    </div>
                    <div
                      class="euiTableCellContent"
                    >
                      <span
                        class="euiTableCellContent__text"
                      >
                        8
                      </span>
                    </div>
                  </td>
                  <td
                    class="euiTableRowCell"
                  >
                    <div
                      class="euiTableRowCell__mobileHeader euiTableRowCell--hideForDesktop"
                    >
                      Imputation method
                    </div>
                    <div
                      class="euiTableCellContent"
                    >
                      <span
                        class="euiTableCellContent__text"
                      >
                        Custom
                      </span>
                    </div>
                  </td>
                  <td
                    class="euiTableRowCell"
                  >
                    <div
                      class="euiTableRowCell__mobileHeader euiTableRowCell--hideForDesktop"
                    >
                      Custom values
                    </div>
                    <div
                      class="euiTableCellContent euiTableCellContent--overflowingContent"
                    >
                      <div
                        class=""
                      >
                        <p>
                          denyMax:5
                        </p>
                        <p>
                          denySum:10
                        </p>
                      </div>
                    </div>
                  </td>
                  <td
                    class="euiTableRowCell"
                  >
                    <div
                      class="euiTableRowCell__mobileHeader euiTableRowCell--hideForDesktop"
                    >
                      Suppression rules
                    </div>
                    <div
                      class="euiTableCellContent euiTableCellContent--overflowingContent"
                    >
                      <div
                        class=""
                      >
                        <button
                          class="euiButtonEmpty euiButtonEmpty--primary euiButtonEmpty--small"
                          type="button"
                        >
                          <span
                            class="euiButtonContent euiButtonEmpty__content"
                          >
                            <span
                              class="euiButtonEmpty__text"
                            >
                              2
                               rules
                            </span>
                          </span>
                        </button>
                      </div>
                    </div>
                  </td>
                </tr>
              </tbody>
            </table>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
